### PR TITLE
Add an hook for newsletter registration/unregistration

### PR DESCRIPTION
**Pull request by jf-viguier**
I've added two hooks :
- actionNewsletterRegistration that can return an error in hookError arg
- displayNewsletterRegistration for the tpl. This hook has to be added to hook/ps_emailsubscription.tpl of classic theme.

Fixes https://github.com/PrestaShop/PrestaShop/issues/17075

### CODE REVIEW

File: ps_emailsubscription.php [hunk 1]
```diff
@@ -395,6 +395,13 @@ public function newsletterRegistration($hookName = null)
             }
         }
 
+        // hook for newsletter registration/unregistration : fill-in hookError string is there is an error
+        $hookError = '';
```
Comment (PierreRambaud) [inline ps_emailsubscription.php]:
```suggestion
        $hookError = null;
```

File: ps_emailsubscription.php [hunk 2]
```diff
@@ -395,6 +395,13 @@ public function newsletterRegistration($hookName = null)
...
+        Hook::exec('actionNewsletterRegistration', ['hookName' => $hookName, 'email' => $_POST['email'], 'action' => $_POST['action'], 'hookError' => &$hookError]);
```
Comment (PierreRambaud) [inline ps_emailsubscription.php]:
```suggestion
        Hook::exec(
            'actionNewsletterRegistration', 
            [
                'hookName' => $hookName, 
                'email' => $_POST['email'], 
                'action' => $_POST['action'], 
                'hookError' => &$hookError,
            ]
        );
```

File: ps_emailsubscription.php [hunk 3]
```diff
@@ -395,6 +395,13 @@ public function newsletterRegistration($hookName = null)
...
+        if ($hookError) {
```
Comment (PierreRambaud) [inline ps_emailsubscription.php]:
```suggestion
        if ($hookError !== null) {
```

File: ps_emailsubscription.php [hunk 4]
```diff
@@ -395,6 +395,21 @@ public function newsletterRegistration($hookName = null)
             }
         }
 
+        // hook for newsletter registration/unregistration : fill-in hookError string is there is an error
+        $hookError = null;
+        Hook::exec(
+            'actionNewsletterRegistration',
```
Comment (PierreRambaud) [inline ps_emailsubscription.php]:
```suggestion
            'actionNewsletterRegistrationBefore', 
```

And could be cool to have the after, and while register / unregister, wdtyt?

**Comment by jeckyl** [comment]
Thank you, this hook becomes mandatory with the implementation of an Ajax treatment so that other modules can continue to use this module to manage subscriptions on third party services.

File: ps_emailsubscription.php [hunk 5]
```diff
@@ -452,6 +467,16 @@ public function newsletterRegistration($hookName = null)
...
             }
         }
+        // hook 
+        Hook::exec(
+            'actionNewsletterRegistrationAfter', 
+            [
+                'hookName' => $hookName, 
+                'email' => $_POST['email'], 
+                'action' => $_POST['action'], 
+                'error' => $this->error,
```
Comment (PierreRambaud) [inline ps_emailsubscription.php]: Should be?

```suggestion
                'error' => &$this->error,
```

**Comment by jf-viguier** [comment]
ok, but actually, is there is an error, return is made before

**Comment by jf-viguier** [comment]
Any news from QA ?

**Comment by sarahdib** [comment]
Hello @jf-viguier 

Thank you for your contribution.

The two hooks aren't available in positions pages is it normal ?
Capture d’écran 2020-04-09 à 11 49 43
Capture d’écran 2020-04-09 à 11 50 03

**Comment by Progi1984** [comment]
Hello @jf-viguier , thanks for your contribution, would you mind to rebase your pull request? Thanks

**Comment by jf-viguier** [comment]
@sarahdib the hooks are created only if a module link itself to one of this hooks.
Otherwise, the hooks has to be hadded in install scripts and config file. Can you help ?

**Comment by sarahdib** [comment]
@jf-viguier Do you have a module you can provide me who do that in order to test the pr ?

**Comment by jf-viguier** [comment]
@Progi1984 I've done a rebase but I don't understand this git nightmare, tell me

**Comment by sarahdib** [comment]
Hello @jf-viguier 

Could you send me a module to test ?

Thank you :)

**Comment by jf-viguier** [comment]
@sarahdib I don't have a test module for this. But it's just a new hook.

**Comment by SimonGrn** [comment]
Hello,
sadly we don't have the time or resources to create modules to test. We need all the informations and resources to be able to test it in time, that's why it's been left behind a little.
@PrestaShop/prestashop-core-developers is it normal that we don't see this hook until a module plugs itself on it ?

**Comment by jf-viguier** [comment]
@SimonGrn yes, it's normal, no need for core team.

**Comment by jf-viguier** [comment]
It's just new hooks, you juste have to test that the code doesn't break something.
IMO Hook usage will be made by modules developers, it's not your concern.

**Comment by Matt75** [comment]
@jf-viguier They need to test if hook works as attended, can you provide a test module ?

You can find examples here : 
- https://github.com/Matt75/testhookdisplayadminproductsextra
- https://github.com/Matt75/testhookcmscategories

Goal is a simple test module hooked on your new hook to demonstrate it works as attended.
You can use PrestaShopLogger to demonstrate it like I do for [testhookdisplayadminproductsextra](https://github.com/Matt75/testhookdisplayadminproductsextra/blob/master/testhookdisplayadminproductsextra.php#L69)

If not this PR will not be approved because cannot be tested.

Thanks for understanding

**Comment by jf-viguier** [comment]
@Matt75 I've done a test module here [https://www.creabilis.com/tmp/crea_testmodule.zip](https://www.creabilis.com/tmp/crea_testmodule.zip)
I hope this will speed up your work.

**Comment by NeOMakinG** [comment]
@sarahdib should be testable now :)

Thanks @jf-viguier

**Comment by jf-viguier** [comment]
The displayNewsletterRegistration hook has to be added to classic ps_emailsubscription.tpl 
The ps_emailsubscription.tpl of the module is overrided by classic theme.

**Comment by jf-viguier** [comment]
I've done a PR for this https://github.com/PrestaShop/PrestaShop/pull/19538

**Comment by sarahdib** [comment]
@jf-viguier Thank you :) It's OK now 
Capture d’écran 2020-06-10 à 11 31 01

Capture d’écran 2020-06-10 à 11 30 53

**Comment by Progi1984** [comment]
Thanks @sarahdib & @jf-viguier

**Pull request merged**